### PR TITLE
GAWB-2647 Simple positive case pet switcheroo test for Sam

### PR DESCRIPTION
**Pull request by jmthibault79**
- Split AuthToken into UserAuthToken / ServiceAccountAuthToken

Tests and depends on https://github.com/broadinstitute/sam/pull/48

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: If you changed a URL that is used elsewhere (e.g. in an email), comment about where it is used and ensure the dependent code is updated.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] **Submitter**: If you're adding new automated UI tests, review the test plan with QA
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Submitter**: Verify all tests go green, including CI tests and automated UI tests.
- [ ] **Submitter**: Squash commits and merge to develop. If adding test code, merge application code and test code at the same time.
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: automation/project/Dependencies.scala
Comment (jmthibault79) [inline automation/project/Dependencies.scala:8]: `ServiceAccountAuthToken` needs these

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/FireCloudClient.scala
Comment (jmthibault79) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/FireCloudClient.scala]: Parses simple JSON case classes

TODO: Jackson equivalents of our Spray formatters

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/Orchestration.scala
Comment (jmthibault79) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/Orchestration.scala:5]: I moved `AuthToken` and its 2 new implementations to a new `auth` package

File: automation/project/Dependencies.scala
Comment (breilly2) [inline automation/project/Dependencies.scala:9]: What exactly is workbench-model needed for? With this pulled in, we have to be careful to not use code that clients don't have (such as serialization) that would make the tests stay green even if we make a breaking change to the request/response format.

File: automation/project/Dependencies.scala
Comment (breilly2) [inline automation/project/Dependencies.scala:41]: Shouldn't workbench-google pull this in as a transitive dependency? Or are you having to intervene to get the right version?

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/ServiceAccountAuthToken.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/ServiceAccountAuthToken.scala]: I'm kind of amazed and pleased that this enhancement fit in so cleanly!

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/analysis/MethodConfigSpec.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/analysis/MethodConfigSpec.scala]: I kinda wish `Credentials` had a function like:
```
  def makeAuthToken(): AuthToken = UserAuthToken(this)
```
so that client code wouldn't have to care about how it's obtained. If it had been there, the number of changes in this PR would have been reduced. Adding it now would be roughly the same number of changes now but fewer changes later (likely never) but would also be cleaner. I leave it up to you whether or not to do it.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/SamApiSpec.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/SamApiSpec.scala]: Does this mean that the conditions of the test are different the second time it's run? For a FiaB instance? Forever?

**Review by breilly2 — COMMENTED** [review]
Minor questions but I'd like to know before I approve. Thanks.

File: automation/project/Dependencies.scala
Comment (rtitle) [reply automation/project/Dependencies.scala:9]: Probably because `GoogleIamDAO` (in workbench-google) uses the `WorkbenchIdentity` case classes in workbench-model. Good point about not using workbench-libs JSON serializers in test code -- I don't think we'd be using the workbench-model JSON serializers in this case, but we could make sure we _don't_ import `WorkbenchIdentityJsonSupport` in any tests.

File: automation/project/Dependencies.scala
Comment (rtitle) [reply automation/project/Dependencies.scala:41]: It's marked as `provided` in workbench-google, so downstream systems must pull it in explicitly.

Actually I'm confused why it's `provided` here too -- who is actually providing the dependency?

(or is it not provided at all, and this is just to make the build work?)

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/FireCloudClient.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/api/FireCloudClient.scala]: One of these imports isn't needed

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/ServiceAccountAuthToken.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/ServiceAccountAuthToken.scala]: Do the tests ever clean up old SA keys when they're done?

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/UserAuthToken.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/UserAuthToken.scala]: This comment isn't really true is it? Looks like it's using an existing pem file.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/ServiceAccountAuthToken.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/ServiceAccountAuthToken.scala]: Maybe could say "a new service account private key" to be more precise

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/SamApiSpec.scala
Comment (rtitle) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/SamApiSpec.scala]: The Sam API does a "get or create". So the first one creates the pet in Google and in LDAP. The second call should return the created pet, not create a new one.

LDAP is local to the fiab, but are the pets removed in Google on test tear down?

**Comment by rtitle** [comment]
A few inline comments, looks good as an initial test case.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/SamApiSpec.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/SamApiSpec.scala]: We're intending this to move to the Sam repo eventually, right? I wonder if would make sense to name the package:
```
org.broadinstitute.dsde.firecloud.test.api.sam
```
so it's clear this is a Sam test. Similarly for Leonardo, Rawls, etc API tests. It might make future refactoring easier.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/ServiceAccountAuthToken.scala
Comment (breilly2) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/ServiceAccountAuthToken.scala]: I think I wrote that comment originally so that's my fault. It could be that I got this wrong. I'd like to understand this better so I can use the right terminology going forward.

**Comment by breilly2** [comment]
Also, since this depends on a particular sam branch that has a different name than this one, you should run the system integration tests manually. I've constructed this URL for you that should configure everything properly (but please double-check my work): https://fc-jenkins.dsp-techops.broadinstitute.org/job/automated-fiab-tests/parambuild/?GOOGLE_PROJ=broad-dsde-qa&firecloud_ui_img=jt_pet_switch&sam_img=vk-gawb-2647&BUILD_SERVICES=true&MESSAGE=manual%20run%20for%20firecloud-ui%20PR%201048

File: automation/project/Dependencies.scala
Comment (jmthibault79) [reply automation/project/Dependencies.scala:41]: I haven't thoroughly investigated this.  Without this line, I get a runtime error: `spray/json/JsonWriter` class not found

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/UserAuthToken.scala
Comment (jmthibault79) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/UserAuthToken.scala]: `AuthToken.makeToken()` calls `cred.refreshToken()` - so maybe the comment belongs with `AuthToken` instead.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/SamApiSpec.scala
Comment (jmthibault79) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/SamApiSpec.scala]: sure!

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/ServiceAccountAuthToken.scala
Comment (jmthibault79) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/ServiceAccountAuthToken.scala]: Um.  Yeah I should fix that.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/analysis/MethodConfigSpec.scala
Comment (jmthibault79) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/analysis/MethodConfigSpec.scala]: sure

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/SamApiSpec.scala
Comment (jmthibault79) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/SamApiSpec.scala]: I'll add a pet cleaning service

Comment (breilly2) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/SamApiSpec.scala]: Based on my experience with writing a registration test, I recommend cleaning your pet before the test as well as after just in case it wasn't cleaned up correctly for some reason. Also, can you fail the test if the pet isn't clean? Otherwise the test could succeed but not really be testing what it claims to.

Comment (jmthibault79) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/SamApiSpec.scala]: Here's the plan:
* ask Google if the pet exists -> fail if true
* create pet
* ask Google if the Pet exists -> fail if false
* [more tests]
* remove pet
* ask Google if the pet exists -> fail if true

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/ServiceAccountAuthToken.scala
Comment (rtitle) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/ServiceAccountAuthToken.scala]: Ok - I think if you're deleting the pet SAs themselves the keys should be deleted along with the pets.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/SamApiSpec.scala
Comment (rtitle) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/SamApiSpec.scala]: Sounds reasonable. You should be able to use `GoogleIamDAO. findServiceAccount/removeServiceAccount` for this.

**Comment by jmthibault79** [comment]
New commits based on comments.  A few pieces remain:
* I have not been able to figure out what is going on with `JsonWriter`.  How concerned are we about this?
* waiting on a new pet deletion Sam endpoint to help clean up

**Comment by jmthibault79** [comment]
Added cleanup using the Sam pet delete endpoint from https://github.com/broadinstitute/sam/pull/51

File: automation/project/Dependencies.scala
Comment (rtitle) [reply automation/project/Dependencies.scala:41]: ok, fine to leave as is then

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/sam/SamApiSpec.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/sam/SamApiSpec.scala]: I think Sam delete pet should remove it from Google:
https://github.com/broadinstitute/sam/blob/develop/src/main/scala/org/broadinstitute/dsde/workbench/sam/service/UserService.scala#L185

Did you find that not to be the case?

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/sam/SamApiSpec.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/sam/SamApiSpec.scala:63]: I think deleting the pet also removes all keys associated with that SA, but this doesn't hurt.

**Review by rtitle — APPROVED** [review]
Looks good, and build is green!

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/sam/SamApiSpec.scala
Comment (jmthibault79) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/sam/SamApiSpec.scala]: Huh.  Yes I did find it necessary to explicitly remove the SA from Google.

Comment (jmthibault79) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/sam/SamApiSpec.scala]: I'm having success without it now, so I'm going to take it out.

Comment (rtitle) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/sam/SamApiSpec.scala]: ok cool

Comment (jmthibault79) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/sam/SamApiSpec.scala]: (nope, I take that back)

Comment (jmthibault79) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/sam/SamApiSpec.scala]: https://broadinstitute.atlassian.net/browse/GAWB-2867

Comment (rtitle) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/api/sam/SamApiSpec.scala]: thank you

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/ServiceAccountAuthToken.scala
Comment (ansingh7115) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/ServiceAccountAuthToken.scala:21]: await.result -> futureValue 
🍰 :D

Comment (jmthibault79) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/auth/ServiceAccountAuthToken.scala:21]: Yep.  But I'm gonna do it in the other PR.

**Review by ansingh7115 — APPROVED** [review]
looks great! 
Most of my concerns were in the `jt_sam_sa` branch

**Pull request merged**
